File: electron/exporter.js
'use strict'
const path = require('path')
const fs = require('fs')
const AdmZip = require('adm-zip')
const sharpPool = require('./modules/sharpPool')
const { sanitizeFilename: sanitize } = require('./utils')

/**
 * CBZ / EPUB 导出器
 *
 * CBZ 格式：zip 打包图片，保持命名顺序
 * CBZ 是基于 Zip 的漫画书格式（Comic Book Archive），
 * 可用任何支持 CBZ 的阅读器打开（YACReader, Kavita, ComicRack, Kindle 等）
 */
function detectImageFormat(buffer) {
  if (buffer[0] === 0xFF && buffer[1] === 0xD8 && buffer[2] === 0xFF) return 'jpg'
  if (buffer[0] === 0x89 && buffer[1] === 0x50 && buffer[2] === 0x4E && buffer[3] === 0x47) return 'png'
  if (buffer[0] === 0x47 && buffer[1] === 0x49 && buffer[2] === 0x46) return 'gif'
  if (buffer[8] === 0x57 && buffer[9] === 0x45 && buffer[10] === 0x42 && buffer[11] === 0x50) return 'webp'
  return 'webp'
}

class ArchiveExporter {
  /**
   * 从下载好的漫画目录导出 CBZ
   * @param {object} opts
   * @param {string} opts.sourceDir     - 下载好的漫画目录（含章节子目录）
   * @param {string} opts.outputPath    - 输出 .cbz 文件路径
   * @param {string} opts.title         - 漫画标题
   * @param {Array<{name:string, dir:string}>} opts.chapters - 章节信息（按下载顺序）
   * @param {function} opts.onProgress  - 进度回调
   * @returns {string} 输出路径
   */
  async toCBZ(opts) {
    const { sourceDir, outputPath, title, chapters, onProgress } = opts
    const zip = new AdmZip()
    let fileCount = 0

    // 如果有封面
    const coverPath = path.join(sourceDir, 'cover.webp')
    if (fs.existsSync(coverPath)) {
      zip.addFile('cover.webp', fs.readFileSync(coverPath))
      fileCount++
    }

    // 遍历章节目录
    for (let i = 0; i < chapters.length; i++) {
      const ch = chapters[i]
      const chDir = (ch.dir && path.isAbsolute(ch.dir))
        ? ch.dir
        : path.join(sourceDir, ch.dir || `${i + 1}-${sanitize(ch.name)}`)
      if (!fs.existsSync(chDir)) {
        console.warn(`[CBZ] 章节目录不存在，跳过: ${chDir}`)
        continue
      }

      const files = fs.readdirSync(chDir)
        .filter(f => /\.(webp|jpg|jpeg|png|gif)$/i.test(f))
        .sort()

      for (let j = 0; j < files.length; j++) {
        const filePath = path.join(chDir, files[j])
        // 用 3 位数字前缀 + 章节序号保证全局排序正确
        // 格式: ch${章节序号三维}_${图片序号三维}.webp
        const zipName = `ch${String(i + 1).padStart(3, '0')}_${String(j + 1).padStart(4, '0')}.webp`
        zip.addFile(zipName, fs.readFileSync(filePath))
        fileCount++
      }

      if (onProgress) {
        onProgress({ current: i + 1, total: chapters.length, fileCount })
      }
    }

    // 写文件
    zip.writeZip(outputPath)
    console.log(`[CBZ] 导出完成: ${outputPath} (${fileCount} 张图片, ${(fs.statSync(outputPath).size / 1024 / 1024).toFixed(1)} MB)`)

    // 写元数据
    this._writeComicInfo(outputPath, { title, chapters: chapters.length })

    return outputPath
  }

  /**
   * 从下载好的漫画目录导出 EPUB
   * 支持分卷：每卷包含指定数量的章节
   * 支持图片质量压缩
   * 生成与参考文件完全一致的结构：
   * - mimetype
   * - META-INF/container.xml
   * - EPUB/css/style.css
   * - EPUB/nav.xhtml
   * - EPUB/toc.ncx
   * - EPUB/content.opf
   * - EPUB/images/{章节文件夹名}_{序号}.webp
   * - EPUB/chapters/{章节文件夹名}.xhtml
   * @param {object} opts
   * @param {string} opts.sourceDir     - 下载好的漫画目录
   * @param {string} opts.outputPath    - 输出 .epub 文件路径（分卷时会自动添加 _1, _2 后缀）
   * @param {string} opts.title         - 漫画标题
   * @param {object} [opts.meta]        - 元数据
   * @param {Array}  opts.chapters      - 章节信息
   * @param {function} opts.onProgress  - 进度回调
   * @param {number} [opts.chaptersPerVolume] - 每卷章节数，undefined/null 表示不分卷
   * @param {string} [opts.imageQuality] - 图片质量：'original' | 'high' | 'medium'
   * @returns {string|string[]} 不分卷返回路径，分卷返回路径数组
   */
  async toEPUB(opts) {
    const { sourceDir, outputPath, title, chapters, onProgress, meta, chaptersPerVolume, imageQuality } = opts

    // 不分卷：直接生成单个 EPUB
    if (!chaptersPerVolume || chaptersPerVolume <= 0 || chapters.length <= chaptersPerVolume) {
      return await this._generateSingleEPUB({ sourceDir, outputPath, title, chapters, onProgress, meta })
    }

    // 分卷：按 chaptersPerVolume 拆分章节
    const volumeCount = Math.ceil(chapters.length / chaptersPerVolume)
    const outputPaths = []
    const outputDir = path.dirname(outputPath)
    const outputBaseName = path.basename(outputPath, '.epub')

    for (let vol = 0; vol < volumeCount; vol++) {
      const startIdx = vol * chaptersPerVolume
      const endIdx = Math.min(startIdx + chaptersPerVolume, chapters.length)
      const volumeChapters = chapters.slice(startIdx, endIdx)
      const volumeTitle = volumeCount === 1 ? title : `${title}_${vol + 1}`
      const volumePath = path.join(outputDir, `${outputBaseName}_${vol + 1}.epub`)

      console.log(`[EPUB] 生成第 ${vol + 1}/${volumeCount} 卷，章节 ${startIdx + 1}-${endIdx}`)

      const result = await this._generateSingleEPUB({
        sourceDir,
        outputPath: volumePath,
        title: volumeTitle,
        chapters: volumeChapters,
        onProgress: (p) => {
          if (onProgress) {
            const totalProgress = (vol * chaptersPerVolume + p.current) / chapters.length * 100
            onProgress({ current: p.current, total: p.total, volume: vol + 1, volumeCount, progress: totalProgress })
          }
        },
        meta,
        imageQuality
      })
      outputPaths.push(result)
    }

    console.log(`[EPUB] 分卷完成，共 ${volumeCount} 卷: ${outputPaths.join(', ')}`)
    return outputPaths
  }

  /**
   * 生成单个 EPUB 文件（内部方法）
   */
  async _generateSingleEPUB(opts) {
    const { sourceDir, outputPath, title, chapters, onProgress, meta, imageQuality } = opts
    const zip = new AdmZip()

    // EPUB 必须的文件结构
    zip.addFile('mimetype', Buffer.from('application/epub+zip'))
    zip.addFile('META-INF/container.xml', Buffer.from(this._containerXML()))

    // 收集 manifest 条目
    const manifestItems = []
    // 收集 NCX navPoints
    const navPoints = []

    // ---------- 1. CSS ----------
    zip.addFile('EPUB/css/style.css', Buffer.from(this._cssContent()))
    manifestItems.push({ id: 'css', href: 'css/style.css', mediaType: 'text/css' })

    // ---------- 2. 封面 ----------
    const coverPath = path.join(sourceDir, 'cover.webp')
    let coverAdded = false
    if (fs.existsSync(coverPath)) {
      let coverBuf = fs.readFileSync(coverPath)
      let coverFormat = detectImageFormat(coverBuf)
      if (coverFormat !== 'webp') {
        try {
          coverBuf = await sharpPool.webpConvertToBuffer(coverBuf, { quality: 85 })
          coverFormat = 'webp'
        } catch (e) {
          console.warn('[EPUB] 封面WebP转换失败，保留原格式', e.message)
        }
      }
      const coverMediaType = { 'webp': 'image/webp', 'jpg': 'image/jpeg', 'jpeg': 'image/jpeg', 'png': 'image/png', 'gif': 'image/gif' }[coverFormat] || 'image/webp'
      zip.addFile('EPUB/images/cover.webp', coverBuf)
      manifestItems.push({ id: 'cover', href: 'images/cover.webp', mediaType: coverMediaType })
      coverAdded = true
    }

    // ---------- 3. 收集所有章节图片 ----------
    const allChapters = []
    for (let i = 0; i < chapters.length; i++) {
      const ch = chapters[i]
      const chDir = ch.dir && path.isAbsolute(ch.dir)
        ? ch.dir
        : path.join(sourceDir, ch.dir || `${ch.index + 1 || i + 1}-${sanitize(ch.name)}`)
      if (!fs.existsSync(chDir)) continue

      const files = fs.readdirSync(chDir)
        .filter(f => /\.(webp|jpg|jpeg|png|gif)$/i.test(f))
        .sort((a, b) => {
          const na = parseInt(a.match(/^(\d+)/)?.[1] || '0', 10)
          const nb = parseInt(b.match(/^(\d+)/)?.[1] || '0', 10)
          return na - nb
        })

      if (files.length === 0) continue

      const chFolderName = ch.dir ? path.basename(ch.dir) : `${ch.index + 1 || i + 1}-${sanitize(ch.name)}`

      const imgNames = []
      for (let j = 0; j < files.length; j++) {
        const srcPath = path.join(chDir, files[j])
        const imgFileName = `${chFolderName}_${String(j + 1).padStart(3, '0')}.webp`

        let imgBuffer = fs.readFileSync(srcPath)
        let actualFormat = detectImageFormat(imgBuffer)

        if (actualFormat !== 'webp') {
          try {
            imgBuffer = await sharpPool.webpConvertToBuffer(imgBuffer, { quality: 85 })
            actualFormat = 'webp'
          } catch (e) {
            console.warn(`[EPUB] 图片WebP转换失败，保留原格式: ${files[j]}`, e.message)
          }
        }

        if (imageQuality && imageQuality !== 'original') {
          try {
            const quality = imageQuality === 'high' ? 80 : 50
            const maxWidth = imageQuality === 'high' ? 1200 : 800
            imgBuffer = await sharpPool.resizeWebpToBuffer(imgBuffer, { maxWidth, quality })
            actualFormat = 'webp'
          } catch (e) {
            console.warn(`[EPUB] 图片压缩失败，使用原图: ${files[j]}`, e.message)
          }
        }

        const mediaType = { 'webp': 'image/webp', 'jpg': 'image/jpeg', 'jpeg': 'image/jpeg', 'png': 'image/png', 'gif': 'image/gif' }[actualFormat] || 'image/webp'

        zip.addFile(`EPUB/images/${imgFileName}`, imgBuffer)
        manifestItems.push({ id: `img-${i}-${j}`, href: `images/${imgFileName}`, mediaType })
        imgNames.push(imgFileName)
      }

      allChapters.push({ chFolderName, chName: ch.name || chFolderName, imgNames })

      if (onProgress) onProgress({ current: i + 1, total: chapters.length })
    }

    // ---------- 4. 单个 XHTML（所有章节合并，锚点分隔） ----------
    const xhtmlId = 'all-pages'
    zip.addFile('EPUB/chapters/all-pages.xhtml', Buffer.from(this._allPagesXHTML(allChapters)))
    manifestItems.push({ id: xhtmlId, href: 'chapters/all-pages.xhtml', mediaType: 'application/xhtml+xml' })

    for (const ch of allChapters) {
      navPoints.push({
        id: ch.chFolderName,
        label: ch.chFolderName,
        src: `chapters/all-pages.xhtml#${ch.chFolderName}`
      })
    }

    // ---------- 4. nav.xhtml ----------
    manifestItems.push({ id: 'nav', href: 'nav.xhtml', mediaType: 'application/xhtml+xml', properties: 'nav' })
    zip.addFile('EPUB/nav.xhtml', Buffer.from(this._navXHTML(title, navPoints)))

    // ---------- 5. toc.ncx ----------
    manifestItems.push({ id: 'ncx', href: 'toc.ncx', mediaType: 'application/x-dtbncx+xml' })
    const isbn = `comic_${sanitize(title)}`
    zip.addFile('EPUB/toc.ncx', Buffer.from(this._ncxXML({ title, isbn, navPoints })))

    // ---------- 6. content.opf ----------
    const opfXml = this._opfXML({
      title: meta?.title || title || 'Untitled',
      author: meta?.author || 'Unknown',
      language: meta?.language || 'zh-CN',
      isbn,
      manifestItems,
      navPoints,
      coverAdded
    })
    zip.addFile('EPUB/content.opf', Buffer.from(opfXml))

    // 写入文件
    zip.writeZip(outputPath)
    console.log(`[EPUB] 导出完成: ${outputPath} (${(fs.statSync(outputPath).size / 1024 / 1024).toFixed(1)} MB)`)
    return outputPath
  }

  /**
   * 写 CBZ 的 ComicInfo.xml 元数据
   */
  _writeComicInfo(cbzPath, meta) {
    const xml = `<?xml version="1.0" encoding="utf-8"?>
<ComicInfo xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
  <Title>${escHtml(meta.title || '')}</Title>
  <PageCount>${meta.fileCount || 0}</PageCount>
</ComicInfo>`
    try {
      const zip = new AdmZip(cbzPath)
      zip.addFile('ComicInfo.xml', Buffer.from(xml))
      zip.writeZip(cbzPath)
    } catch (e) {
      console.warn('[CBZ] 写入 ComicInfo.xml 失败:', e.message)
    }
  }

  _containerXML() {
    return `<?xml version="1.0" encoding="UTF-8"?>
<container version="1.0" xmlns="urn:oasis:names:tc:opendocument:xmlns:container">
  <rootfiles>
    <rootfile full-path="EPUB/content.opf" media-type="application/oebps-package+xml"/>
  </rootfiles>
</container>`
  }

  _cssContent() {
    return `* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}
html, body {
    width: 100%;
    line-height: 0;
}
.comic-container {
    width: 100%;
}
.no-images {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    font-size: 18px;
    color: #666;
    text-align: center;
    background-color: #f5f5f5;
}
.comic-img {
    display: block !important;
    width: 100% !important;
    height: auto !important;
    margin: 0 !important;
    padding: 0 !important;
    border: none !important;
}
.chapter-nav {
    position: fixed;
    bottom: 20px;
    width: 100%;
    display: flex;
    justify-content: space-between;
    padding: 0 20px;
    z-index: 999;
}
.nav-btn {
    padding: 10px 20px;
    background: #333;
    color: #fff;
    border: none;
    border-radius: 5px;
    cursor: pointer;
}
.nav-btn:disabled {
    background: #999;
    cursor: not-allowed;
}
@media (-webkit-min-device-pixel-ratio: 2), (min-resolution: 192dpi) {
    .comic-img {
        image-rendering: crisp-edges;
    }
}`
  }

  /**
   * EPUB3 导航文档（nav.xhtml）
   */
  _navXHTML(title, navPoints) {
    const items = navPoints.map(p =>
      `        <li>\n          <a href="${escHtml(p.src)}">${escHtml(p.label)}</a>\n        </li>`
    ).join('\n')
    return `<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" lang="zh-CN" xml:lang="zh-CN">
  <head>
    <title>${escHtml(title)}</title>
  </head>
  <body>
    <nav epub:type="toc" id="id" role="doc-toc">
      <h2>${escHtml(title)}</h2>
      <ol>
${items}
      </ol>
    </nav>
  </body>
</html>`
  }

  _chapterXHTML(title, imgNames) {
    const imgs = imgNames.map(n =>
      `        <img src="../images/${n}" alt="${escHtml(title)}" class="comic-img"/>`
    ).join('\n')
    return `<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="zh-CN" xml:lang="zh-CN">
  <head>
    <title>${escHtml(title)}</title>
    <link rel="stylesheet" type="text/css" href="../css/style.css"/>
  </head>
  <body>
    <div class="comic-container">
${imgs}
    </div>
  </body>
</html>`
  }

  _allPagesXHTML(allChapters) {
    const sections = allChapters.map(ch => {
      const anchor = `<a id="${escHtml(ch.chFolderName)}"></a>`
      const imgs = ch.imgNames.map(n =>
        `        <img src="../images/${n}" alt="${escHtml(ch.chName)}" class="comic-img"/>`
      ).join('\n')
      return `      ${anchor}\n${imgs}`
    }).join('\n')
    return `<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="zh-CN" xml:lang="zh-CN">
  <head>
    <title>Comic</title>
    <link rel="stylesheet" type="text/css" href="../css/style.css"/>
  </head>
  <body>
    <div class="comic-container">
${sections}
    </div>
  </body>
</html>`
  }

  /**
   * 生成 content.opf
   */
  _opfXML(opts) {
    const { title, author, language, isbn, manifestItems, navPoints, coverAdded } = opts

    const manifestXml = manifestItems.map(item => {
      const props = item.properties ? ` properties="${escHtml(item.properties)}"` : ''
      return `        <item id="${escHtml(item.id)}" href="${escHtml(item.href)}" media-type="${escHtml(item.mediaType)}"${props}/>`
    }).join('\n')

    const xhtmlItems = manifestItems.filter(item => item.mediaType === 'application/xhtml+xml' && item.id !== 'nav')
    const spineXml = xhtmlItems.map(item =>
      `        <itemref idref="${escHtml(item.id)}"/>`
    ).join('\n')

    // metadata
    let metadataXml = `        <dc:identifier id="bookid">${escHtml(isbn)}</dc:identifier>\n`
    metadataXml += `        <dc:title>${escHtml(title)}</dc:title>\n`
    if (author && author !== 'Unknown') {
      metadataXml += `        <dc:creator>${escHtml(author)}</dc:creator>\n`
    }
    metadataXml += `        <dc:language>${escHtml(language || 'zh-CN')}</dc:language>`

    return `<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://www.idpf.org/2007/opf" unique-identifier="bookid" version="3.0">
    <metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
${metadataXml}
    </metadata>
    <manifest>
${manifestXml}
    </manifest>
    <spine toc="ncx">
${spineXml}
    </spine>
</package>`
  }

  /**
   * 生成 toc.ncx
   */
  _ncxXML(opts) {
    const { title, isbn, navPoints } = opts
    const navPointsXml = navPoints.map(p =>
      `        <navPoint id="${escHtml(p.id)}">\n            <navLabel>\n                <text>${escHtml(p.label)}</text>\n            </navLabel>\n            <content src="${escHtml(p.src)}"/>\n        </navPoint>`
    ).join('\n')

    return `<?xml version="1.0" encoding="utf-8"?>
<ncx xmlns="http://www.daisy.org/z3986/2005/ncx/" version="2005-1">
    <head>
        <meta content="${escHtml(isbn)}" name="dtb:uid"/>
        <meta content="0" name="dtb:depth"/>
        <meta content="0" name="dtb:totalPageCount"/>
        <meta content="0" name="dtb:maxPageNumber"/>
    </head>
    <docTitle>
        <text>${escHtml(title)}</text>
    </docTitle>
    <navMap>
${navPointsXml}
    </navMap>
</ncx>`
  }
}

function escHtml(s) { return String(s).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;') }

module.exports = new ArchiveExporter()

File: src-tauri/src/exporter.rs
// EPUB / CBZ 导出器（移植 comic-app/electron/exporter.js）
// CBZ：zip 打包图片，命名 ch{章:03}_{页:04}.webp 保证全局顺序
// EPUB：mimetype + META-INF/container.xml + EPUB/{css,images,chapters,nav,toc,content.opf}
use crate::paths;
use std::io::Write;
use std::path::{Path, PathBuf};
use zip::write::SimpleFileOptions;

pub struct ExportChapter {
    pub name: String,
    pub dir: PathBuf,
}

fn list_images_sorted(dir: &Path) -> Vec<PathBuf> {
    paths::list_chapter_images(dir)
}

/// 导出 CBZ
pub fn to_cbz(
    source_dir: &Path,
    output_path: &Path,
    chapters: &[ExportChapter],
) -> Result<usize, String> {
    let file = std::fs::File::create(output_path).map_err(|e| e.to_string())?;
    let mut zip = zip::ZipWriter::new(file);
    let opts = SimpleFileOptions::default().compression_method(zip::CompressionMethod::Deflated);
    let mut count = 0usize;

    let cover = source_dir.join("cover.webp");
    if cover.exists() {
        if let Ok(data) = std::fs::read(&cover) {
            zip.start_file("cover.webp", opts).map_err(|e| e.to_string())?;
            zip.write_all(&data).map_err(|e| e.to_string())?;
            count += 1;
        }
    }

    for (i, ch) in chapters.iter().enumerate() {
        if !ch.dir.exists() {
            continue;
        }
        let files = list_images_sorted(&ch.dir);
        for (j, f) in files.iter().enumerate() {
            if let Ok(data) = std::fs::read(f) {
                let name = format!("ch{:03}_{:04}.webp", i + 1, j + 1);
                zip.start_file(name, opts).map_err(|e| e.to_string())?;
                zip.write_all(&data).map_err(|e| e.to_string())?;
                count += 1;
            }
        }
    }
    zip.finish().map_err(|e| e.to_string())?;
    Ok(count)
}

fn xml_escape(s: &str) -> String {
    s.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
}

fn container_xml() -> &'static str {
    r#"<?xml version="1.0" encoding="UTF-8"?>
<container version="1.0" xmlns="urn:oasis:names:tc:opendocument:xmlns:container">
  <rootfiles>
    <rootfile full-path="EPUB/content.opf" media-type="application/oebps-package+xml"/>
  </rootfiles>
</container>"#
}

fn css_content() -> &'static str {
    "html,body{margin:0;padding:0;background:#111;}\nimg{display:block;width:100%;height:auto;margin:0 auto;}\n"
}

/// 导出 EPUB（单卷，所有章节合并到 all-pages.xhtml，锚点分隔）
pub fn to_epub(
    source_dir: &Path,
    output_path: &Path,
    title: &str,
    author: &str,
    chapters: &[ExportChapter],
) -> Result<usize, String> {
    let file = std::fs::File::create(output_path).map_err(|e| e.to_string())?;
    let mut zip = zip::ZipWriter::new(file);
    let stored = SimpleFileOptions::default().compression_method(zip::CompressionMethod::Stored);
    let deflated = SimpleFileOptions::default().compression_method(zip::CompressionMethod::Deflated);

    // mimetype 必须第一个且 stored（不压缩）
    zip.start_file("mimetype", stored).map_err(|e| e.to_string())?;
    zip.write_all(b"application/epub+zip").map_err(|e| e.to_string())?;

    zip.start_file("META-INF/container.xml", deflated).map_err(|e| e.to_string())?;
    zip.write_all(container_xml().as_bytes()).map_err(|e| e.to_string())?;

    zip.start_file("EPUB/css/style.css", deflated).map_err(|e| e.to_string())?;
    zip.write_all(css_content().as_bytes()).map_err(|e| e.to_string())?;

    let mut manifest: Vec<String> = Vec::new();
    let mut spine: Vec<String> = Vec::new();
    let mut nav_points: Vec<(String, String)> = Vec::new();
    let mut img_count = 0usize;

    manifest.push(r#"<item id="css" href="css/style.css" media-type="text/css"/>"#.to_string());

    // 封面
    let mut cover_added = false;
    let cover = source_dir.join("cover.webp");
    if cover.exists() {
        if let Ok(data) = std::fs::read(&cover) {
            zip.start_file("EPUB/images/cover.webp", deflated).map_err(|e| e.to_string())?;
            zip.write_all(&data).map_err(|e| e.to_string())?;
            manifest.push(
                r#"<item id="cover" href="images/cover.webp" media-type="image/webp"/>"#.to_string(),
            );
            cover_added = true;
        }
    }

    // 章节图片 + 合并 xhtml body
    let mut body = String::new();
    for (i, ch) in chapters.iter().enumerate() {
        if !ch.dir.exists() {
            continue;
        }
        let files = list_images_sorted(&ch.dir);
        if files.is_empty() {
            continue;
        }
        let folder = ch
            .dir
            .file_name()
            .map(|s| s.to_string_lossy().to_string())
            .unwrap_or_else(|| format!("{}", i + 1));
        body.push_str(&format!("<div id=\"ch_{}\"></div>\n", i));
        nav_points.push((format!("ch_{}", i), xml_escape(&ch.name)));
        for (j, f) in files.iter().enumerate() {
            if let Ok(data) = std::fs::read(f) {
                let img_name = format!("{}_{:03}.webp", paths::sanitize_filename(&folder), j + 1);
                zip.start_file(format!("EPUB/images/{}", img_name), deflated)
                    .map_err(|e| e.to_string())?;
                zip.write_all(&data).map_err(|e| e.to_string())?;
                manifest.push(format!(
                    r#"<item id="img-{}-{}" href="images/{}" media-type="image/webp"/>"#,
                    i, j, img_name
                ));
                body.push_str(&format!("<img src=\"../images/{}\"/>\n", img_name));
                img_count += 1;
            }
        }
    }

    // all-pages.xhtml
    let xhtml = format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"><head><meta charset="utf-8"/>
<title>{}</title><link rel="stylesheet" type="text/css" href="../css/style.css"/></head>
<body>
{}
</body></html>"#,
        xml_escape(title),
        body
    );
    zip.start_file("EPUB/chapters/all-pages.xhtml", deflated).map_err(|e| e.to_string())?;
    zip.write_all(xhtml.as_bytes()).map_err(|e| e.to_string())?;
    manifest.push(
        r#"<item id="all-pages" href="chapters/all-pages.xhtml" media-type="application/xhtml+xml"/>"#
            .to_string(),
    );
    spine.push(r#"<itemref idref="all-pages"/>"#.to_string());

    // nav.xhtml
    let nav_lis: String = nav_points
        .iter()
        .map(|(id, label)| {
            format!(
                "<li><a href=\"chapters/all-pages.xhtml#{}\">{}</a></li>",
                id, label
            )
        })
        .collect::<Vec<_>>()
        .join("\n");
    let nav = format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops"><head><meta charset="utf-8"/><title>{}</title></head>
<body><nav epub:type="toc"><h1>目录</h1><ol>{}</ol></nav></body></html>"#,
        xml_escape(title),
        nav_lis
    );
    zip.start_file("EPUB/nav.xhtml", deflated).map_err(|e| e.to_string())?;
    zip.write_all(nav.as_bytes()).map_err(|e| e.to_string())?;
    manifest.push(r#"<item id="nav" href="nav.xhtml" media-type="application/xhtml+xml" properties="nav"/>"#.to_string());

    // toc.ncx
    let isbn = format!("comic_{}", paths::sanitize_filename(title));
    let nav_pts: String = nav_points
        .iter()
        .enumerate()
        .map(|(n, (id, label))| {
            format!(
                r#"<navPoint id="np_{}" playOrder="{}"><navLabel><text>{}</text></navLabel><content src="chapters/all-pages.xhtml#{}"/></navPoint>"#,
                n, n + 1, label, id
            )
        })
        .collect::<Vec<_>>()
        .join("\n");
    let ncx = format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<ncx xmlns="http://www.daisy.org/z3986/2005/ncx/" version="2005-1">
<head><meta name="dtb:uid" content="{}"/></head>
<docTitle><text>{}</text></docTitle>
<navMap>{}</navMap></ncx>"#,
        isbn,
        xml_escape(title),
        nav_pts
    );
    zip.start_file("EPUB/toc.ncx", deflated).map_err(|e| e.to_string())?;
    zip.write_all(ncx.as_bytes()).map_err(|e| e.to_string())?;
    manifest.push(r#"<item id="ncx" href="toc.ncx" media-type="application/x-dtbncx+xml"/>"#.to_string());

    // content.opf
    let cover_meta = if cover_added {
        r#"<meta name="cover" content="cover"/>"#
    } else {
        ""
    };
    let opf = format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<package xmlns="http://www.idpf.org/2007/opf" version="3.0" unique-identifier="bookid">
<metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
<dc:identifier id="bookid">{}</dc:identifier>
<dc:title>{}</dc:title>
<dc:creator>{}</dc:creator>
<dc:language>zh-CN</dc:language>
{}
</metadata>
<manifest>
{}
</manifest>
<spine toc="ncx">
{}
</spine>
</package>"#,
        isbn,
        xml_escape(title),
        xml_escape(author),
        cover_meta,
        manifest.join("\n"),
        spine.join("\n")
    );
    zip.start_file("EPUB/content.opf", deflated).map_err(|e| e.to_string())?;
    zip.write_all(opf.as_bytes()).map_err(|e| e.to_string())?;

    zip.finish().map_err(|e| e.to_string())?;
    Ok(img_count)
}
